google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,391 @@
+        self._rules_sema = threading.BoundedSemaphore(value=1)
+
+        self.resource_rules_map = {
+            applies_to: collections.defaultdict(set)
+            for applies_to in SUPPORTED_RETENTION_RES_TYPES}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book
+
+        Args:
+            rule_defs (dict): rule definitions dictionary
+        """
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        self._rules_sema.acquire()
+        try:
+            applies_to = rule_def.get('applies_to')
+            if applies_to is None:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of applies_to in rule {}'.format(rule_index))
+
+            minimum_retention, maximum_retention = get_retention_range(
+                rule_def, rule_index)
+
+            if any(support_type not in SUPPORTED_RETENTION_RES_TYPES
+                   for support_type in applies_to):
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Invalid applies_to resource in rule {}'.format(rule_index))
+
+            added_applies_to = set()
+
+            for appto in applies_to:
+                if appto in added_applies_to:
+                    raise audit_errors.InvalidRulesSchemaError(
+                        'redundant applies_to in rule {}'.format(rule_index))
+                added_applies_to.add(appto)
+
+                self.create_and_add_rule(
+                    rule_def,
+                    rule_index,
+                    appto,
+                    minimum_retention,
+                    maximum_retention)
+        finally:
+            self._rules_sema.release()
+
+    def create_and_add_rule(self, rule_def, rule_index, apply_to,
+                            min_retention, max_retention):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+            apply_to (str): The resource type that the rule is applied to
+            min_retention(int): minimum value of the age in lifecycle
+            max_retention(int): maximum value of the age in lifecycle
+        """
+        resource = rule_def.get('resource', None)
```

Comment:
default for .get is None, no need to put ", None"

Reply:
consider making the default [] and moving the "if resource is None" check above and making it "if 'resource' not in rule_def"